extensions/amp-form/0.1/amp-form.js

```diff
@@ -245,17 +275,109 @@ function onInvalidInputBlur_(event) {
+    input.removeEventListener('keyup', onInvalidInputKeyUp_);
+
+    input.addEventListener('keyup', onInvalidInputKeyUp_);
+    input.addEventListener('blur', onInvalidInputBlur_);
+    validationBubble.show(input, input.validationMessage);
+  }
+}
+
+
+/**
+ * Checks user validity for all inputs, fieldsets and the form.
+ * @param {!HTMLFormElement} form
+ */
+function checkUserValidityOnSubmission_(form) {
+  const inputs = form.querySelectorAll('input,select,textarea,fieldset');
+  for (let i = 0; i < inputs.length; i++) {
+    checkUserValidity(inputs[i]);
+  }
+  checkUserValidity(form);
+}
+
+
+/**
+ * Returns the user validity state of the element.
+ * @param {!HTMLInputElement|!HTMLFormElement|!HTMLFieldSetElement} element
+ * @return {string}
+ */
+function getUserValidityStateFor(element) {
+  if (element.classList.contains('user-valid')) {
+    return UserValidityState.USER_VALID;
+  } else if (element.classList.contains('user-invalid')) {
+    return UserValidityState.USER_INVALID;
+  }
+
+  return UserValidityState.NONE;
+}
 
-  // Remove any previous listeners on the same input. This avoids adding many
-  // listeners on the same element when the user submit pressing Enter or any
-  // other method to submit the form without the element losing focus.
-  input.removeEventListener('blur', onInvalidInputBlur_);
-  input.removeEventListener('keyup', onInvalidInputKeyUp_);
 
-  input.addEventListener('keyup', onInvalidInputKeyUp_);
-  input.addEventListener('blur', onInvalidInputBlur_);
-  validationBubble.show(input, input.validationMessage);
+/**
+ * Checks user validity which applies .user-valid and .user-invalid AFTER the user
+ * interacts with the input by moving away from the input (blur) or by changing its
+ * value (input).
+ *
+ * See :user-invalid spec for more details:
+ *   https://drafts.csswg.org/selectors-4/#user-pseudos
+ *
+ * The specs are still not fully specified. The current solution tries to follow a common
+ * sense approach for when to apply these classes. As the specs gets clearer, we should
+ * strive to match it as much as possible.
+ *
+ * @param {!HTMLInputElement|!HTMLFormElement|!HTMLFieldSetElement} element
+ * @returns {boolean} true to indicate that validity should propagate to ancestors.
+ */
+function checkUserValidity(element) {
+  const previousValidityState = getUserValidityStateFor(element);
+  const isCurrentlyValid = element.checkValidity();
+  if (previousValidityState != UserValidityState.USER_VALID &&
+      isCurrentlyValid) {
+    element.classList.add('user-valid');
+    element.classList.remove('user-invalid');
+    // Don't propagate user-valid if this was unmarked before.
+    return previousValidityState != UserValidityState.NONE;
+  } else if (previousValidityState != UserValidityState.USER_INVALID &&
+      !isCurrentlyValid) {
+    element.classList.add('user-invalid');
+    element.classList.remove('user-valid');
+    // Always propagate an invalid state change.
```

Comment:
Expanded the comment. The moment the input becomes `user-invalid` its ancestors are also `user-invalid`.